lib/routes/account.js

```diff
@@ -1013,41 +1118,77 @@ module.exports = function (
         }
       },
       handler: function (request, reply) {
-        log.begin('Account.RecoveryEmailVerify', request)
         var uid = request.payload.uid
         var code = Buffer(request.payload.code, 'hex')
+
+        log.begin('Account.RecoveryEmailVerify', request)
         db.account(Buffer(uid, 'hex'))
           .then(
             function (account) {
-              // If the account is already verified, they may be e.g.
-              // clicking a stale link.  Silently succeed.
-              if (account.emailVerified) {
-                log.increment('account.already_verified')
-                return true
-              }
-              if (!butil.buffersAreEqual(code, account.emailCode)) {
-                throw error.invalidVerificationCode()
-              }
-              log.timing('account.verified', Date.now() - account.createdAt)
-              log.event('verified', request, {
-                email: account.email,
-                uid: account.uid,
-                locale: account.locale
-              })
-              log.increment('account.verified')
 
-              // send a push notification to all devices that the account changed
-              push.notifyUpdate(uid, 'accountVerify')
+              /**
+               * Logic for account and token verification
+               *
+               * 1) Attempt to use code as tokenVerificationId to verify session.
+               *
+               * 2) An error is thrown if tokenVerificationId does not exist (check to see if email
+               *    verification code) or the tokenVerificationId does not correlate to the
+               *    account uid (damaged linked/spoofed account)
+               *
+               * 3) Verify account email if not already verified.
+               */
+              return db.verifyToken(code, account)
+                .then(function () {
+                  log.event('tokenVerified', request, {
+                    email: account.email,
+                    uid: account.uid,
+                    locale: account.locale,
+                    tokenVerificationId: code
+                  })
+                })
+                .catch(function (err) {
+                  if (err.errno === error.ERRNO.INVALID_TOKEN && butil.buffersAreEqual(code, account.emailCode)) {
```

Comment:
I don't believe we can remove the `butil.buffersAreEqual` because it checks to see if code is an emailCode. Without this check it would verify the account email even if invalid token found. Not sure the reason why `butil.buffersAreEqual` was used instead of `code === emailCode`.